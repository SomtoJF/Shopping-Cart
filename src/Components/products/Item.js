import { Link } from "react-router-dom";
import "../../styles/products/Item.sass";
import { useState } from "react";

const Item = (props) => {
	const product = props.product;
	const breakdown = props.breakdown;
	const setCart = props.setCart;
	const cart = props.cart;
	if (breakdown === true) {
		return <BrokenDownItem item={product} />;
	} else {
		return <ExpandedItem item={product} setCart={setCart} cart={cart} />;
	}
};

export default Item;

const BrokenDownItem = (props) => {
	const product = props.item;
	return (
		<li className="product-item">
			<img src={product.src} alt={product.alt} />
			<details>
				<summary>Product Information</summary>
				<ul>
					<li>
						<div>Product</div>
						<div>{product.alt}</div>
					</li>
					<li>
						<div>Photographer</div>
						<div>{product.photographer}</div>
					</li>
					<li>
						<div>Price</div>
						<div>${product.price}</div>
					</li>
					<li>
						<Link to={"/Shopping-Cart/products/" + product.id}>More Information</Link>
					</li>
				</ul>
			</details>
		</li>
	);
};

const ExpandedItem = (props) => {
	const product = props.item;
	const setCart = props.setCart;
	const cart = props.cart;
	const [quantity, setQuantity] = useState(1);
	const addToCart = (item, quantity) => {
		item = { quantity: quantity, ...item };
		let newCart = cart
			.filter((cartItem) => cartItem.id !== item.id)
			.concat(item);
		setCart(newCart);
	};
	return (
		<main id="a-product">
			<img src={product.src} alt={product.alt} />
			<section>
				<ul>
					<li>
						<div>Product</div>
						<div>{product.alt}</div>
					</li>
					<li>
						<div>Photographer</div>
						<a href={product.photographer_url}>{product.photographer}</a>
					</li>

					<li>
						<div>Price</div>
						<div>${product.price}</div>
					</li>
					<li>
						<div>Poster Size</div>
						<div>{product.poster_size} cm</div>
					</li>
					<li>
						<div>Color</div>
						<div
							style={{
								backgroundColor: `${product.color}`,
							}}
						>
							{product.color}
						</div>
					</li>
					<li>
						<button
							type="button"
							onClick={() => {
								addToCart(product, quantity);
							}}
						>
							Add to Cart
						</button>

						<div id="quantity-controls">
							<button
								type="button"
								onClick={() => {
									setQuantity(quantity - 1);
								}}
								disabled={quantity <= 1 ? true : false}
							>
								-
							</button>
							<span> Qty {quantity} </span>
							<button
								type="button"
								onClick={() => {
									setQuantity(quantity + 1);
								}}
							>
								+
							</button>
						</div>
					</li>
				</ul>
			</section>
		</main>
	);
};
