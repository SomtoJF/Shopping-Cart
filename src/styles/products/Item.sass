@use '../../App'
@import url('https://fonts.googleapis.com/css2?family=Teko&display=swap')

.product-item::marker
    font-size: 7rem
    font-family: App.$navbar-font
    
    

.product-item
    box-sizing: border-box
    width: 100%
    padding: 0px
    margin: 10% 0px
    img
        width: 100%
        margin-top: -15%
        position: relative
        z-index: -1

    details
        summary
            font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif
            font-weight: 600

        ul
            list-style-type: none
            padding: 0px

            li
                display: flex
                justify-content: space-between
                align-items: center
                border-bottom: 1px solid App.$secondary-color
                height: 5vh
                overflow: hidden
                gap: 20px
            
            li>div:first-child
                font-family: system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif
                font-weight: 600

            li>div:nth-child(2)
                font-family: 'Teko', system-ui, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif,sans-serif
                font-size: 1rem
                text-overflow: ellipsis
                overflow: hidden
                white-space: nowrap
