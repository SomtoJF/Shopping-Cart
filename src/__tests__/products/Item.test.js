import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom"; // optional
import userEvent from "@testing-library/user-event";
import Item from "../../Components/products/Item";

describe("Item Component", () => {
	it("displays an item accurately", () => {
		const anItem = {
			src: "www.somto.me/something.jpg",
			alt: "reach for the moon, aim for the stars",
			photographer: "SomtoJF",
		};

		const { container } = render(<Item product={anItem} />, {
			wrapper: MemoryRouter,
		});

		expect(container).toMatchSnapshot();
	});
});
