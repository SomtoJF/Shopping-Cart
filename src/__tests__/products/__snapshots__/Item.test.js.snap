// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Component in 'products' page properly breaksdown/summarizes a product 1`] = `
<div>
  <li
    class="product-item"
  >
    <img
      alt="reach for the moon, aim for the stars"
      src="www.somto.me/something.jpg"
    />
    <details>
      <summary>
        Product Information
      </summary>
      <ul>
        <li>
          <div>
            Product
          </div>
          <div>
            reach for the moon, aim for the stars
          </div>
        </li>
        <li>
          <div>
            Photographer
          </div>
          <div>
            SomtoJF
          </div>
        </li>
        <li>
          <div>
            Price
          </div>
          <div>
            $
            1000
          </div>
        </li>
        <li>
          <a
            href="/products/undefined"
          >
            More Information
          </a>
        </li>
      </ul>
    </details>
  </li>
</div>
`;

exports[`Item Component in the 'product' page expands a product when the breakdown prop is false 1`] = `
<div>
  <main
    id="a-product"
  >
    <img
      alt="reach for the moon, aim for the stars"
      src="www.somto.me/something.jpg"
    />
    <section>
      <ul>
        <li>
          <div>
            Product
          </div>
          <div>
            reach for the moon, aim for the stars
          </div>
        </li>
        <li>
          <div>
            Photographer
          </div>
          <a>
            SomtoJF
          </a>
        </li>
        <li>
          <div>
            Price
          </div>
          <div>
            $
            1000
          </div>
        </li>
        <li>
          <div>
            Poster Size
          </div>
          <div>
            1000 x 1000
             cm
          </div>
        </li>
        <li>
          <div>
            Color
          </div>
          <div
            style="background-color: rgb(255, 255, 255);"
          >
            #FFFFFF
          </div>
        </li>
        <li>
          <button
            type="button"
          >
            Add to Cart
          </button>
          <div
            id="quantity-controls"
          >
            <button
              disabled=""
              type="button"
            >
              -
            </button>
            <span>
               Qty 
              1
               
            </span>
            <button
              type="button"
            >
              +
            </button>
          </div>
        </li>
      </ul>
    </section>
  </main>
</div>
`;
